### ISSUE TITLE
String-to-Atom conversion seems way too slow

### ISSUE THREAD

**Issue by pcwalton** [servo/servo#6906]
We can't reach competitive performance on Dromaeo until we optimize our DOMString-to-Atom conversion. Switching from SipHash to FNV will help some, but I'm sure it won't help enough.

Is Gecko doing something like storing precomputed hashes for DOM strings to avoid hashing them over and over for Atom conversion?

**Comment by bzbarsky** [comment]
Gecko doesn't store precomputed hashes.

However what Gecko _does_ do for a setAttribute call is walk the existing list of attributes looking for one which matches the given name.  See the `InternalGetExistingAttrNameFromQName` call at the top of `Element::SetAttribute`.  This is needed anyway, because you have to find the right namespace/prefix for the given name, etc.

`InternalGetExistingAttrNameFromQName` works on non-interned strings, precisely to avoid the cost of interning.

**Comment by SimonSapin** [comment]
`Atom::from_slice` hashes twice. (Once for the static atom set, once for the dynamic map.) I’m working on a PR to rust-phf to allow re-using the phf hash.

**Comment by fitzgen** [comment]
Here is a profile of `setAttribute` calls in a loop after #8645. I'm still pretty unfamiliar with Servo, but I've added my analysis as well.

<img width="1199" alt="screen shot 2015-11-24 at 10 34 33 am" src="https://cloud.githubusercontent.com/assets/74571/11376691/fdd721bc-9296-11e5-88e4-d99ea9e8a655.png">

* Heaviest function is spending 10.3% of time doing some kind of string manipulation which is called from auto-generated bindings code. I'm not sure how much room for improvement there is here, as I'm totally unfamiliar with bindings generation.

* Next heaviest, 9.7% of time spent validating if the attribute name is valid. I suspect there could be some fast path inserted here for obviously valid attribute names or the most common attribute names.

* Next up is `Element::SetAttribute` itself with 7.8% of time. Unclear to me how much more can be skimmed from here; seems pretty tight as it is. The only work it is directly doing (rather than happening in some callee) is the search described in https://github.com/servo/servo/issues/6906#issuecomment-127332489 and implemented in #8545.

* Finally, there is a bunch of stuff related to bumping the node's version number and possibly marking it dirty. Collectively, all this work is taking up the most time. I wonder how much of this work could be delayed until absolutely required, for example if/when getAttribute is called or on the next reflow or style recalc.

**Comment by bzbarsky** [comment]
> doing some kind of string manipulation

This is basically the "make a copy of the string the JS engine gave us" bit that's more or less required nowadays because we can't trust the JS engine's strings to stay put...  The important thing to check here is that we're not allocating, just copying onto the stack, for short strings.

**Comment by fitzgen** [comment]
It seems a bit more invasive of a change, but: why not properly root + trace strings from SM rather than copy? Was this a fundamental architectural decision or something that just happened?

**Comment by bzbarsky** [comment]
That could be done as long as all downstream consumers can deal with both (for the sake of simplicity let's assume both) SM string representations.  That means branching everything that manipulates the strings in any way on one-byte vs two-byte, right?

**Comment by jdm** [comment]
I'm 99% sure we allocate for all strings.

**Comment by asajeffrey** [comment]
I'm playing around with string representations at the moment. This is interesting data!

Part of the problem with JSStrings is they're UCS2-encoded or Latin1-encoded by SpiderMonkey, whereas Servo uses UTF-8. There has to be some re-encoding, but we may be able to save on some round-trips. In particular, I'd like to avoid a lot of the DOMString to/from Atom conversion.

**Comment by eefriedman** [comment]
> Next heaviest, 9.7% of time spent validating if the attribute name is valid. I suspect there could be some fast path inserted here for obviously valid attribute names or the most common attribute names.

Potential stupid trick: if there's an existing attribute with the given name, it must be valid.  Not sure how useful that would be in practice.

**Comment by bzbarsky** [comment]
For benchmarks, pretty useful, probably.

**Comment by fitzgen** [comment]
> if there's an existing attribute with the given name, it must be valid.

Trying this out.

**Comment by bzbarsky** [comment]
> if there's an existing attribute with the given name, it must be valid.

By the way, it's not obvious that this is true.  Last I checked, the set of names `setAttribute` considers valid per spec is not the same as the set of names of attributes the HTML parser can create.  There was some call for aligning the two, but not sure whether it happened.  @annevk, do you recall offhand?

**Comment by eefriedman** [comment]
> Finally, there is a bunch of stuff related to bumping the node's version number and possibly marking it dirty. Collectively, all this work is taking up the most time. I wonder how much of this work could be delayed until absolutely required, for example if/when getAttribute is called or on the next reflow or style recalc.

It's probably possible to speed some of this up... attribute_mutated checks the attribute name in a way which looks pretty inefficient.

See #8227 for some context on DOM versioning.

**Comment by asajeffrey** [comment]
The problem is that we need some way to invalidate the caches for live HTMLCollections. Rather annoyingly, in most cases this is wasted effort, but any way of detecting when work is being wasted is likely to take longer than updating the version numbers.

**Comment by bzbarsky** [comment]
> The problem is that we need some way to invalidate the caches for live HTMLCollections.

That only needs to happen if the value actually changed, right?

**Comment by asajeffrey** [comment]
Yes, we don't need to dirty the node if the old and new values are the same. (I don't know how much this check would buy us though.)

**Comment by annevk** [comment]
@bzbarsky I think https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/YyqxpIPU-0E is the latest. Perhaps @foolip or @domenic know more about Chromium internal efforts to align `createElement`/`setAttribute` et al with the HTML parser.

**Comment by foolip** [comment]
I didn't investigate any further after that thread ended. If I remember correctly, the last idea on the table was to remove all restrictions on what characters can appear in an attribute or element name, and if necessary to do some kind of coercion when serializing instead.

If that's something that seems worth doing, maybe open a DOM spec bug and I can add telemetry to see how often the existing validation causes exceptions to be thrown? (With methods this common, it's not implausible that stuff actually depends on the exceptions being thrown.)

**Comment by fitzgen** [comment]
See also https://groups.google.com/forum/#!topic/mozilla.dev.servo/iY7JRKwfCIM

> I've created a [crate](https://github.com/fitzgen/inlinable_string) that provides a mostly drop-in replacement of
std::string::String but with inline storage for small strings to avoid heap
allocation. It also provides a StringExt trait to abstract string
operations over both InlinableString and std::string::String. I'm planning
on implementing this trait for DOMString as well, making the codegen emit
bindings code that uses InlinableString for stack allocated strings, and
make the webidl interface methods use `StringExt` instead of `DOMString`
directy. My hope is that this will make a good chunk of that performance
bottleneck go away.

Planning on experimenting with this soon.

**Comment by fitzgen** [comment]
> I'm playing around with string representations at the moment.

@asajeffrey, somehow I missed the comment that you were playing with string representations :-/

 What are you trying out? Can you comment on how it compares to https://github.com/servo/servo/issues/6906#issuecomment-161152810? Any initial findings?

**Comment by asajeffrey** [comment]
My code is at https://github.com/asajeffrey/servo/tree/string_representation_experiments, though it's not the cleanest. I'll have a look at your work tomorrow!

**Comment by fitzgen** [comment]
> My code is at https://github.com/asajeffrey/servo/tree/string_representation_experiments, though it's not the cleanest. I'll have a look at your work tomorrow!

I see you modified the `DOMString` type directly, not just the stack allocated strings in emitted binding code (which was my plan). Although not optimal because it increases the memory footprint of `DOMString`, that's a lot easier to implement, and I was able to mimic the approach. I have a branch that makes `DOMString` use `InlinableString` rather than `std::string::String` now: https://github.com/servo/servo/compare/master...fitzgen:domstring-use-inlinable-string

Frustratingly, I haven't observed any performance wins on either dromaeo nor my micro-benchmark despite seeing that 10% reported time pushing strings disappear from profiler results and not seeing any obvious replacement related to `InlinableString` pop up.

**Comment by asajeffrey** [comment]
Yes, welcome to my world of implementing different string representations and seeing no benefit :)

I think what's causing this is that the source of memory allocation for DOMStrings is in clone() and converting to/from JSString. I instrumented these methods, and saw that there's almost no use of clone(), but lots of conversions of tiny strings. Looking at the conversion code:

https://github.com/servo/servo/blob/master/components/util/str.rs#L133
https://github.com/servo/rust-mozjs/blob/master/src/conversions.rs#L389
https://github.com/servo/rust-mozjs/blob/master/src/conversions.rs#L401

you can see that a heap-allocated buffer is being created in each case, which means jemalloc is being thrashed even if the DOMString is stack-allocated.

**Comment by asajeffrey** [comment]
Looking at your branch, I can see you fixed the buffer in str.rs, but not in rust-mozjs/src/conversions.rs. Since many strings are latin-1, this may explain what's going on.

**Comment by asajeffrey** [comment]
I'm currently experimenting with removing the buffering for converting
short strings to/from JS. We shall see if this improves matters...

**Comment by asajeffrey** [comment]
My attempt at removing buffering: https://github.com/asajeffrey/servo/commit/1533238eb8ffff22688572a04d5bd6eb8ac88297, which got me no measurable benefit.

**Comment by asajeffrey** [comment]
Another attempt, this time using horrible bit-twiddling to pack a representation that's either an Atom, a String, or an inlined string into 3 words. Still no joy. https://github.com/asajeffrey/servo/tree/domstring-use-atom-or-inlined-string

**Comment by nox** [comment]
Could this be checked again now that we have `impl From<String> for Atom` and we use it in `Element::SetAttribute`?

**Comment by samuknet** [comment]
Hi,

In making some experiments for Servo, I have found that `document.createElement` can take roughly up to 3-4x longer in Servo compared to FF/Chrome.  This is significantly slowing down quite a few of the experiments I am creating to showcase Servo.    Could this be the same issue as the one documented here? If needed, I can file a separate issue for this.

Thanks.

**Comment by bzbarsky** [comment]
Almost certainly a separate issue.  Needs a testcase and profiling.
